src/osmet/mod.rs

```diff
@@ -0,0 +1,579 @@
+// Copyright 2020 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::fs::{File, OpenOptions};
+use std::io::{self, Seek, SeekFrom, Write};
+use std::os::unix::fs::FileTypeExt;
+use std::path::{Path, PathBuf};
+
+use error_chain::bail;
+use nix::mount;
+use serde::{Deserialize, Serialize};
+use walkdir::WalkDir;
+use xz2::write::XzEncoder;
+
+use crate::blockdev::*;
+use crate::cmdline::*;
+use crate::errors::*;
+
+mod fiemap;
+mod file;
+mod io_helpers;
+mod unpacker;
+
+use crate::osmet::fiemap::*;
+use crate::osmet::io_helpers::*;
+use crate::osmet::file::*;
+use crate::osmet::unpacker::*;
+
+// just re-export OsmetUnpacker
+pub use crate::osmet::unpacker::OsmetUnpacker;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Mapping {
+    extent: Extent,
+    object: Sha256Digest,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct OsmetPartition {
+    start_offset: u64,
+    end_offset: u64,
+    mappings: Vec<Mapping>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Osmet {
+    /// Partitions for which we've registered mappings.
+    partitions: Vec<OsmetPartition>,
+    /// Checksum of the final disk image. Used for unpacking verification.
+    checksum: Sha256Digest,
+    /// Size of the final disk image. Used for unpacking verification.
+    size: u64,
+    /// Required sector size of target block device during unpacking.
+    sector_size: u32,
+}
```

Comment:
Would it make sense to include the contents of `/etc/os-release`?

Reply:
Can you expand on this? More for making inspecting an osmet file by hand more palatable, or for some kind of sanity-checking?

Reply:
I was originally thinking of an osmet file as a longer-lived artifact, and thought we might want it to be more self-describing.  Since that isn't the goal, I don't think it's necessary to have a machine-readable summary of the image.

However, there's still https://github.com/coreos/coreos-installer/pull/197/files#r402686439.  One approach could be to add a `description` header field, containing a string like `Fedora CoreOS 31.20200323.2.1 x86_64, 512-byte sectors`.  It'd be cleaner to have `osmet pack` take that from a command-line arg, rather than parsing `os-release` itself.

Or we could just encode all those details in the osmet filename and print that from `Display`.